File: tests/interop/websocket/py_node/test_utils.py
import json
import time
from typing import Any

import trio


class ResultCollector:
    def __init__(self) -> None:
        self.results: dict[str, dict[str, Any]] = {}
        self.errors: list[str] = []
        self.start_time = time.time()

    def add_result(self, test_name: str, success: bool, details: Any = None) -> None:
        self.results[test_name] = {
            "success": success,
            "details": details,
            "timestamp": time.time(),
            "duration": time.time() - self.start_time,
        }

    def add_error(self, error: str) -> None:
        self.errors.append(str(error))

    def to_dict(self) -> dict[str, Any]:
        return {
            "results": self.results,
            "errors": self.errors,
            "total_tests": len(self.results),
            "passed": sum(1 for r in self.results.values() if r["success"]),
            "failed": sum(1 for r in self.results.values() if not r["success"]),
            "total_duration": time.time() - self.start_time,
        }


async def wait_for_server_ready(host: str, port: int, timeout: float = 10.0) -> bool:
    import socket

    end_time = time.time() + timeout
    while time.time() < end_time:
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.settimeout(1.0)
            result = sock.connect_ex((host, port))
            sock.close()
            if result == 0:
                return True
        except Exception:
            pass
        await trio.sleep(0.5)
    return False


def save_results_to_file(
    results: dict[str, Any], filename: str = "test_results.json"
) -> None:
    try:
        with open(filename, "w") as f:
            json.dump(results, f, indent=2, default=str)
    except Exception:
        pass


def TestResults() -> ResultCollector:
    return ResultCollector()

File: extra/py-libp2p/tests/interop/websocket/js_node/test_utils.js
export class TestResults {
    constructor() {
        this.results = {}
        this.errors = []
        this.startTime = Date.now()
    }

    addResult(testName, success, details = null) {
        this.results[testName] = {
            success,
            details,
            timestamp: Date.now(),
            duration: Date.now() - this.startTime
        }
    }

    addError(error) {
        this.errors.push(error.toString())
    }

    toJSON() {
        return {
            results: this.results,
            errors: this.errors,
            totalTests: Object.keys(this.results).length,
            passed: Object.values(this.results).filter(r => r.success).length,
            failed: Object.values(this.results).filter(r => !r.success).length,
            totalDuration: Date.now() - this.startTime
        }
    }

    printSummary() {
        const data = this.toJSON()
        console.log('\n' + '='.repeat(50))
        console.log('TEST RESULTS SUMMARY')
        console.log('='.repeat(50))
        console.log(`Total Tests: ${data.totalTests}`)
        console.log(`Passed: ${data.passed}`)
        console.log(`Failed: ${data.failed}`)
        console.log(`Duration: ${data.totalDuration}ms`)

        if (this.errors.length > 0) {
            console.log(`\nErrors (${this.errors.length}):`)
            this.errors.forEach(error => console.log(`  - ${error}`))
        }

        console.log('\nDetailed Results:')
        Object.entries(this.results).forEach(([testName, result]) => {
            const status = result.success ? '✓ PASS' : '✗ FAIL'
            console.log(`  ${testName}: ${status} (${result.duration}ms)`)
            if (result.details && !result.success) {
                console.log(`    Details: ${JSON.stringify(result.details)}`)
            }
        })
    }
}

export async function waitForServerReady(host, port, timeout = 10000) {
    const startTime = Date.now()

    while (Date.now() - startTime < timeout) {
        try {
            const response = await fetch(`http://${host}:${port}`, {
                method: 'HEAD',
                signal: AbortSignal.timeout(1000)
            })
            return true
        } catch (error) {
            await new Promise(resolve => setTimeout(resolve, 500))
        }
    }

    return false
}

export async function saveResultsToFile(results, filename = 'test_results.json') {
    try {
        const fs = await import('fs')
        fs.writeFileSync(filename, JSON.stringify(results, null, 2))
        console.log(`Results saved to ${filename}`)
    } catch (error) {
        console.error(`Failed to save results: ${error}`)
    }
}
